### PR TITLE
add Speech commands dataset

### LINKED ISSUE #3283: Add Speech Commands dataset
## Adding a Dataset
- **Name:** Speech commands
- **Description:** A Dataset for Limited-Vocabulary Speech Recognition
- **Paper:** https://arxiv.org/abs/1804.03209
- **Data:** https://www.tensorflow.org/datasets/catalog/speech_commands,  Available:
http://download.tensorflow.org/data/speech_commands_v0.02.tar.gz
- **Motivation:** Nice dataset for audio classification training

cc @anton-l 

Instructions to add a new dataset can be found [here](https://github.com/huggingface/datasets/blob/master/ADD_NEW_DATASET.md).

### PR DESCRIPTION
**Pull request by polinaeterna**
closes  #3283

### CODE REVIEW

**Comment by polinaeterna** [comment]
OMG I hope it's more or less clear what I mean...

**Comment by polinaeterna** [comment]
Closes  #3283

**Comment by polinaeterna** [comment]
@anton-l ping

File: datasets/speech_commands/speech_commands.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+            name="v0.01",
+            description=textwrap.dedent(
+                """\
+                Version 0.01 of SpeechCommands dataset. Contains twenty command words:
+                digits from zero to nine, "Yes", "No", "Up", "Down", "Left", "Right", "On", "Off", "Stop", and "Go";
+                several auxiliary words (label `_unknown_`), and background noise (label `_background_noise_`)
+                """
+            ),
+            labels=LABELS_V1,
+            version=datasets.Version("0.0.1"),
```
Comment (anton-l) [inline datasets/speech_commands/speech_commands.py:129]: I think it's rather "0.1.0"?

File: datasets/speech_commands/speech_commands.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        SpeechCommandsConfig(
+            name="v0.02",
+            description=textwrap.dedent(
+                """\
+                Version 0.02 of SpeechCommands dataset.
+                Contains more command and auxiliary words than version 0.01.
+                """
+            ),
+            labels=LABELS_V2,
+            version=datasets.Version("0.0.2"),
```
Comment (anton-l) [inline datasets/speech_commands/speech_commands.py:140]: Same here

File: datasets/speech_commands/speech_commands.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+            ),
+            labels=LABELS_V1,
+            version=datasets.Version("0.0.1"),
+        ),
+        SpeechCommandsConfig(
+            name="v0.02",
+            description=textwrap.dedent(
+                """\
+                Version 0.02 of SpeechCommands dataset.
+                Contains more command and auxiliary words than version 0.01.
```
Comment (anton-l) [inline datasets/speech_commands/speech_commands.py:136]:
```suggestion
                Version 0.02 of the SpeechCommands dataset.
                Contains more commands and auxiliary words than version 0.01.
```

File: datasets/speech_commands/speech_commands.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+                    "speaker_id": None,
+                    "utterance_id": 0,
+                }
+                continue
+
+            elif word in self.config.labels[:-2]:  # the last two labels are _unknown_ and _background_
+                label = word
+            else:
+                label = UNKNOWN
+                # TODO: or maybe should I preserve words outside the WORDS list too and
```
Comment (anton-l) [inline datasets/speech_commands/speech_commands.py:204]: Having the `full_label` field to keep the original words would be ideal, yes!

File: datasets/speech_commands/speech_commands.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+of them five times.: "Yes", "No", "Up", "Down", "Left", "Right", "On", "Off", "Stop", "Go".
+More command words were added in the version 0.02:  "Zero", "One", "Two", "Three", "Four", "Five", "Six",
+"Seven", "Eight", "Nine", "Backward",  "Forward", "Follow", "Learn".
+
+To help distinguish unrecognized words, there are also ten auxiliary words ( label `_unknown_`),
+which most speakers only said once. These include "Bed", "Bird", "Cat", "Dog", "Happy", "House", "Marvin", "Sheila",
+"Tree", and "Wow".
+
+There is also a `_background_noise_` class containin a set of  longer audio clips that are either recordings or
+a mathematical simulations of noise.
```
Comment (anton-l) [inline datasets/speech_commands/speech_commands.py:64]:
```suggestion
There is also a `_background_noise_` class containing a set of longer audio clips that are either recordings or
a mathematical simulation of noise.
```

File: datasets/speech_commands/speech_commands.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
+# coding=utf-8
+# Copyright 2020 The HuggingFace Datasets Authors and the current dataset script contributor.
```
Comment (anton-l) [inline datasets/speech_commands/speech_commands.py:2]:
```suggestion
# Copyright 2021 The HuggingFace Datasets Authors and the current dataset script contributor.
```

File: datasets/speech_commands/README.md [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+into a predefined set of words. The task is usually performed on-device for the 
+fast response time. Thus, accuracy, model size, and inference time are all crucial.
+
+### Languages
+
+The language data in SpeechCommands is in English (BCP-47 `en`).
+
+## Dataset Structure
+
+### Data Instances
```
Comment (anton-l) [inline datasets/speech_commands/README.md:85]: Thanks for including the examples for different groups of lables!

File: datasets/speech_commands/README.md [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+- crowdsourced
+languages:
+- en
+licenses:
+- cc-by-4.0
+multilinguality:
+- monolingual
+pretty_name: SpeechCommands
+size_categories:
+- unknown
```
Comment (anton-l) [inline datasets/speech_commands/README.md:14]:
```suggestion
  v0.01:
    - 10K<n<100K
  v0.02:
    - 100K<n<1M
```

File: datasets/speech_commands/README.md [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+source_datasets:
+- original
+task_categories:
+- speech-processing
+task_ids:
+- keyword-spotting
```
Comment (anton-l) [inline datasets/speech_commands/README.md:20]:
```suggestion
- other-other-keyword-spotting
```

**Review by anton-l — CHANGES_REQUESTED** [review]
Thanks for working on this dataset, Polina! Feel free to convert this PR to review-ready once you add the auxiliary label field, the integration looks clean overall :slightly_smiling_face:

**Comment by polinaeterna** [comment]
@lhoestq 
Hi Quentin! Thank you for your feedback and suggestions! 🤗

Yes, that was actually what I wanted to do next - I mean the steaming stuff :)
Also, I need to make some changes to the readme (to account for the updated features set).

Hopefully, I will be done by tomorrow afternoon if that's ok.

**Comment by polinaeterna** [comment]
@lhoestq  Hi Quentin!

I've implemented (hopefully, correctly) the streaming compatibility but the problem with the current approach is that we first need to iterate over the full archive anyway to get the list of filenames for train and validation sets (see [this](https://github.com/huggingface/datasets/pull/3335/files#diff-aeea540d136025e30a842856779e9c6485a5dc6fc9eb7fd6d3be2acd2f49b8e3R186), the same approach is implemented in TFDS version). Only after that, we can generate examples, so we cannot stream the dataset before the first iteration ends and it takes some time. It's probably not the most effective way. 

If the streaming mode is turned off, this approach (with two iterations) is actually slower than the previous implementation (with archive extraction). 

My suggestion is to host separate archives for each split prepared in advance. That way there would be no need for iterating over the common archive to collect train and validation filenames.  @anton-l  suggested to make AWS mirrors for them. I've prepared these archives, for now you can take a look at them [here](https://drive.google.com/drive/u/0/folders/1oMrZHzPgHAKprKJuvih91CM8KMSzh_pL). I simplified their structure a bit so if we switch to using them, the code then should be changed (and simplified) a bit too.

**Comment by lhoestq** [comment]
Hi ! Thanks for the changes :)

> My suggestion is to host separate archives for each split prepared in advance. That way there would be no need for iterating over the common archive to collect train and validation filenames. @anton-l suggested to make AWS mirrors for them. I've prepared these archives, for now you can take a look at them here. I simplified their structure a bit so if we switch to using them, the code then should be changed (and simplified) a bit too.

I agree, I just uploaded them on AWS

https://s3.amazonaws.com/datasets.huggingface.co/SpeechCommands/v0.01/v0.01_test.tar.gz
https://s3.amazonaws.com/datasets.huggingface.co/SpeechCommands/v0.01/v0.01_train.tar.gz
https://s3.amazonaws.com/datasets.huggingface.co/SpeechCommands/v0.01/v0.01_validation.tar.gz
https://s3.amazonaws.com/datasets.huggingface.co/SpeechCommands/v0.02/v0.02_test.tar.gz
https://s3.amazonaws.com/datasets.huggingface.co/SpeechCommands/v0.02/v0.02_validation.tar.gz

Note that in the future we can move those files to actual repositories on the Hugging Face Hub, since we are migrating the datasets from this repository to the Hugging Face Hub (as mirrors), to make them more accessible to the community.

**Comment by polinaeterna** [comment]
@lhoestq Thank you! Gonna look at this tomorrow :)

**Comment by polinaeterna** [comment]
@lhoestq I've modified the code to fit new data format, now it works for v0.01 but doesn't work for v0.02 as the training archive is missing. Could you please create a mirror for that one too? You can find it [here](https://drive.google.com/file/d/1mPjnVMYb-VhPprGlOX8v9TBT1GT-rtcp/view?usp=sharing)

And when it's done I'll need to regenerate all the meta / dummy stuff, and this version will be ready for a review :)

**Comment by lhoestq** [comment]
Here you go :)
https://s3.amazonaws.com/datasets.huggingface.co/SpeechCommands/v0.02/v0.02_train.tar.gz

**Comment by lhoestq** [comment]
FYI I juste merged a fix for the Windows CI error on `master`, feel free to merge `master` again into your branch

**Comment by lhoestq** [comment]
All green ! I had to fix some minor stuff in the CI but it's good now

Next step is to mark it as ready for review, and I think it's all good so we can merge 🚀

**Comment by polinaeterna** [comment]
@lhoestq 🤗

**Review by anton-l — APPROVED** [review]
Everything looks great now, will use it in the classification examples very soon! :slightly_smiling_face:

**Review by lhoestq — APPROVED** [review]
Awesome thank you :)

**Comment by LysandreJik** [comment]
:tada:
